File: js/src/engine/moves.ts
// Legal move generation for the in-browser Hive engine -- a direct port
// of the Python package's engine/moves.py. See that file's docstring for
// the overall design (single-BFS `slideReachable` instead of validating
// one candidate at a time, placement actions keyed by piece *type* not
// instance). Behavior must stay identical to the Python engine, which is
// itself validated against the reference Django app -- see the plan doc's
// Phase 5 (parity fixtures) for how this file gets checked against it.

import { HEX_DIRS, PieceType, TURN_NUMBER_QUEEN_MUST_BE_PLACED } from "./constants.js";
import type { GameState, Owner, Pos } from "./state.js";
import { neighbors, parsePosKey, positionsConnected, posKey, sharedNeighbors } from "./state.js";

export enum MoveKind {
  PLACE = 0,
  MOVE = 1,
  THROW = 2,
}

export interface Move {
  readonly kind: MoveKind;
  readonly pieceId: number; // placed/moved piece, or the pillbug itself for THROW
  readonly to: Pos;
  readonly thrownPieceId?: number; // THROW only: the piece being lifted
}

export function generateLegalMoves(state: GameState): Move[] {
  if (state.gameOver) return [];
  const moves = placementMoves(state);
  if (state.queenPlaced[state.currentPlayer]) {
    moves.push(...boardMoves(state));
    moves.push(...pillbugThrowMoves(state));
  }
  return moves;
}

// --- placement -------------------------------------------------------------

function placementMoves(state: GameState): Move[] {
  const owner = state.currentPlayer;
  const hand = state.hand[owner];
  if (hand.length === 0) return [];
  const positions = placementPositions(state, owner);
  if (positions.size === 0) return [];
  const mustPlaceQueen =
    !state.queenPlaced[owner] && state.turnNo === TURN_NUMBER_QUEEN_MUST_BE_PLACED;

  const moves: Move[] = [];
  const seenTypes = new Set<PieceType>();
  for (const pieceId of hand) {
    const pieceType = state.pieces.get(pieceId)!.pieceType;
    if (seenTypes.has(pieceType)) continue;
    if (mustPlaceQueen && pieceType !== PieceType.QUEEN) continue;
    seenTypes.add(pieceType);
    for (const posKeyStr of positions) {
      moves.push({ kind: MoveKind.PLACE, pieceId, to: parsePosKey(posKeyStr) });
    }
  }
  return moves;
}

function placementPositions(state: GameState, owner: Owner): Set<string> {
  const occupied = state.occupied();
  if (occupied.size === 0) return new Set([posKey([0, 0, 0])]);

  const opponent: Owner = owner === 0 ? 1 : 0;
  if (state.turnNo === 1) {
    // Each side's very first placement just needs to touch the hive so
    // far, which (on turn 1) can only be the opponent's lone piece.
    const anchors = state.boardPositions(opponent);
    const result = new Set<string>();
    for (const anchorKey of anchors) {
      for (const nb of neighbors(parsePosKey(anchorKey))) {
        const key = posKey(nb);
        if (!occupied.has(key)) result.add(key);
      }
    }
    return result;
  }

  const ownPositions = state.boardPositions(owner);
  const oppPositions = state.boardPositions(opponent);
  const candidates = new Set<string>();
  for (const ownKey of ownPositions) {
    for (const nb of neighbors(parsePosKey(ownKey))) {
      const key = posKey(nb);
      if (!occupied.has(key)) candidates.add(key);
    }
  }
  const result = new Set<string>();
  for (const candidateKey of candidates) {
    const touchesOpponent = neighbors(parsePosKey(candidateKey)).some((nb) =>
      oppPositions.has(posKey(nb)),
    );
    if (!touchesOpponent) result.add(candidateKey);
  }
  return result;
}

// --- on-board moves ----------------------------------------------------------

function boardMoves(state: GameState): Move[] {
  const owner = state.currentPlayer;
  const moves: Move[] = [];
  for (const pieceId of state.piecesOnBoard(owner)) {
    const pos = state.position.get(pieceId)!;
    if (state.topPieceAt(pos) !== pieceId) continue; // buried, can't move
    const pieceType = state.pieces.get(pieceId)!.pieceType;
    for (const destKey of destinationsFor(state, pieceId, pieceType)) {
      const dest = parsePosKey(destKey);
      if (moveKeepsHiveConnected(state, pieceId, dest)) {
        moves.push({ kind: MoveKind.MOVE, pieceId, to: dest });
      }
    }
  }
  return moves;
}

function destinationsFor(state: GameState, pieceId: number, pieceType: PieceType): Set<string> {
  switch (pieceType) {
    case PieceType.QUEEN:
    case PieceType.PILLBUG:
      return slideDestinations(state, pieceId, 1);
    case PieceType.ANT:
      return slideDestinations(state, pieceId, null);
    case PieceType.SPIDER:
      return slideDestinations(state, pieceId, 3, 3);
    case PieceType.GRASSHOPPER:
      return grasshopperDestinations(state, pieceId);
    case PieceType.BEETLE:
      return beetleDestinations(state, pieceId);
    case PieceType.LADYBUG:
      return ladybugDestinations(state, pieceId);
    case PieceType.MOSQUITO:
      return mosquitoDestinations(state, pieceId);
    default:
      throw new Error(`unhandled piece type ${pieceType as number}`);
  }
}

/** Occupied-position set after hypothetically moving `pieceId` to `dest`.
 * Only vacates the start cell if this piece was its sole occupant -- a
 * piece stacked under a beetle/mosquito keeps that position "occupied"
 * for connectivity purposes when the top piece leaves. */
function resultingOccupied(state: GameState, pieceId: number, dest: Pos): Set<string> {
  const start = state.position.get(pieceId)!;
  const occ = new Set(state.occupied());
  if (state.stackHeightAt(start) === 1) occ.delete(posKey(start));
  occ.add(posKey(dest));
  return occ;
}

/** The official "One Hive Rule" requires connectivity to hold both
 * *during* a move (the instant the piece is lifted, before it lands --
 * real Hive rules forbid a move that would momentarily split the hive,
 * even if the piece's destination would reconnect it) and *after* it (the
 * final board). Only matters "during" when the start cell truly empties
 * (stack height 1); a beetle/mosquito climbing off a taller stack never
 * affects connectivity by leaving, since something remains there. */
function moveKeepsHiveConnected(state: GameState, pieceId: number, dest: Pos): boolean {
  const start = state.position.get(pieceId)!;
  if (state.stackHeightAt(start) === 1) {
    const during = new Set(state.occupied());
    during.delete(posKey(start));
    if (!positionsConnected(during)) return false;
  }
  return positionsConnected(resultingOccupied(state, pieceId, dest));
}

function slideDestinations(
  state: GameState,
  pieceId: number,
  maxSteps: number | null,
  exactSteps: number | null = null,
): Set<string> {
  const start = state.position.get(pieceId)!;
  const occupiedWithoutStart = new Set(state.occupied());
  occupiedWithoutStart.delete(posKey(start));
  return slideReachable(occupiedWithoutStart, start, maxSteps, exactSteps);
}

/** All destinations reachable from `start` by sliding across `occupied`
 * (which must NOT include `start` itself -- the piece is lifted first).
 * Exact-step sliding (spider) needs a different search than
 * unbounded/max-step sliding (ant/queen) -- see `slideReachableExact` for
 * why -- so this just dispatches to whichever is correct. */
function slideReachable(
  occupied: ReadonlySet<string>,
  start: Pos,
  maxSteps: number | null,
  exactSteps: number | null,
): Set<string> {
  if (exactSteps !== null) return slideReachableExact(occupied, start, exactSteps);

  // A single BFS pass computes every reachable destination instead of
  // validating one candidate at a time: since the wedge/gate check and
  // the "hive stays connected" check at each step only depend on the
  // current node and the single candidate next node (not on the path
  // taken to get there), and *any* non-self-crossing path to a node
  // within maxSteps is equally good enough (there's no "must be exactly
  // this many steps" requirement here), a node's reachability doesn't
  // depend on which path first discovered it -- so a single shared
  // `visited` set across the whole search is exact, not an
  // approximation. (This reasoning does NOT extend to exact-step sliding
  // -- see `slideReachableExact`.)
  const visited = new Set<string>([posKey(start)]);
  const reachable = new Set<string>();
  const queue: [Pos, number][] = [[start, 0]];
  let head = 0;
  while (head < queue.length) {
    const [current, steps] = queue[head++];
    for (const nb of neighbors(current)) {
      const nbKey = posKey(nb);
      if (visited.has(nbKey)) continue;
      visited.add(nbKey);
      if (occupied.has(nbKey)) continue;
      if (!positionsConnected(new Set([...occupied, nbKey]))) continue;
      const [n1, n2] = sharedNeighbors(current, nb);
      if (occupied.has(posKey(n1)) && occupied.has(posKey(n2))) continue; // wedged

      const newSteps = steps + 1;
      if (maxSteps === null || newSteps <= maxSteps) reachable.add(nbKey);
      if (maxSteps === null || newSteps < maxSteps) queue.push([nb, newSteps]);
    }
  }
  return reachable;
}

/** Destinations reachable from `start` in *exactly* `exactSteps` slides
 * (spider), without ever crossing a hex already visited earlier in that
 * same slide.
 *
 * Unlike max-step sliding, this can't reuse one shared `visited` set
 * across the whole search: a hex first reached via a *shorter* path
 * would then block rediscovering it via a *different*, longer,
 * non-self-crossing path that legitimately reaches it in exactly
 * `exactSteps` -- "no revisiting a hex" is a per-path rule, not a global
 * one across every path from `start`. A real game found this concretely:
 * a destination the shared-visited BFS marked unreachable (some other
 * path had already visited it after 2 steps) but which a valid 3-step
 * path -- through different intermediate hexes -- does reach. So this
 * instead explores every non-self-crossing path explicitly; cheap in
 * practice since `exactSteps` is always 3 (only spider uses this) with
 * branching factor at most 6.
 *
 * This exact bug is present in the Python engine's vendored reference
 * oracle's own `can_slide_path` (tests/reference/helpers.py, in the
 * hive-bot repo -- an unmodified copy of the real ttbg-web-app's own
 * code), so it isn't just a bug this port could have introduced -- it's
 * a real, reachable bug in the shipped Django app itself. See
 * engine/moves.py's `_slide_reachable_exact` (the Python side of this
 * same fix) for the full writeup. */
function slideReachableExact(
  occupied: ReadonlySet<string>,
  start: Pos,
  exactSteps: number,
): Set<string> {
  const reachable = new Set<string>();

  function dfs(current: Pos, path: ReadonlySet<string>, steps: number): void {
    if (steps === exactSteps) {
      reachable.add(posKey(current));
      return;
    }
    for (const nb of neighbors(current)) {
      const nbKey = posKey(nb);
      if (path.has(nbKey) || occupied.has(nbKey)) continue;
      if (!positionsConnected(new Set([...occupied, nbKey]))) continue;
      const [n1, n2] = sharedNeighbors(current, nb);
      if (occupied.has(posKey(n1)) && occupied.has(posKey(n2))) continue; // wedged
      dfs(nb, new Set([...path, nbKey]), steps + 1);
    }
  }

  dfs(start, new Set([posKey(start)]), 0);
  return reachable;
}

function grasshopperDestinations(state: GameState, pieceId: number): Set<string> {
  const start = state.position.get(pieceId)!;
  const occupied = state.occupied();
  const destinations = new Set<string>();
  for (const [dq, dr, ds] of HEX_DIRS) {
    let [q, r, s] = start;
    let hoppedOver = 0;
    for (;;) {
      q += dq;
      r += dr;
      s += ds;
      const key = posKey([q, r, s]);
      if (!occupied.has(key)) {
        if (hoppedOver > 0) destinations.add(key);
        break;
      }
      hoppedOver++;
    }
  }
  return destinations;
}

function beetleDestinations(state: GameState, pieceId: number): Set<string> {
  const start = state.position.get(pieceId)!;
  const pieceLevel = state.stackIndexOf(pieceId);
  const destinations = new Set<string>();
  for (const end of neighbors(start)) {
    const endStackLen = state.stackHeightAt(end);
    if (pieceLevel !== endStackLen) {
      // Climbing on/off a stack always changes level -- no gate check.
      destinations.add(posKey(end));
      continue;
    }
    const [n1, n2] = sharedNeighbors(start, end);
    const h1 = state.stackHeightAt(n1) - 1; // top-of-stack index, -1 if empty
    const h2 = state.stackHeightAt(n2) - 1;
    if (h1 >= pieceLevel && h2 >= pieceLevel) continue; // wedged between two taller (or equal) neighbors
    destinations.add(posKey(end));
  }
  return destinations;
}

function ladybugDestinations(state: GameState, pieceId: number): Set<string> {
  const start = state.position.get(pieceId)!;
  const occupied = new Set(state.occupied());
  occupied.delete(posKey(start));
  const destinations = new Set<string>();
  for (const a of neighbors(start)) {
    if (!occupied.has(posKey(a))) continue;
    for (const b of neighbors(a)) {
      if (posKey(b) === posKey(start) || !occupied.has(posKey(b))) continue;
      for (const c of neighbors(b)) {
        const cKey = posKey(c);
        if (cKey === posKey(start) || cKey === posKey(a) || occupied.has(cKey)) continue;
        destinations.add(cKey);
      }
    }
  }
  return destinations;
}

function mosquitoDestinations(state: GameState, pieceId: number): Set<string> {
  if (state.stackIndexOf(pieceId) > 0) {
    // Having climbed up (by copying a beetle) it now only moves as one.
    return beetleDestinations(state, pieceId);
  }
  const start = state.position.get(pieceId)!;
  const adjacentTypes = new Set<PieceType>();
  for (const nb of neighbors(start)) {
    const top = state.topPieceAt(nb);
    if (top !== null) adjacentTypes.add(state.pieces.get(top)!.pieceType);
  }
  adjacentTypes.delete(PieceType.MOSQUITO);
  const destinations = new Set<string>();
  for (const pieceType of adjacentTypes) {
    for (const key of destinationsFor(state, pieceId, pieceType)) destinations.add(key);
  }
  return destinations;
}

// --- pillbug throw -----------------------------------------------------------

function pillbugThrowMoves(state: GameState): Move[] {
  const owner = state.currentPlayer;
  const moves: Move[] = [];
  for (const pillbugId of state.piecesOnBoard(owner)) {
    if (state.pieces.get(pillbugId)!.pieceType !== PieceType.PILLBUG) continue;
    const pos = state.position.get(pillbugId)!;
    if (state.topPieceAt(pos) !== pillbugId) continue;
    const occupied = state.occupied();
    const emptyDests = neighbors(pos).filter((nb) => !occupied.has(posKey(nb)));
    if (emptyDests.length === 0) continue;

    for (const targetPos of neighbors(pos)) {
      const targetId = state.topPieceAt(targetPos);
      if (targetId === null || targetId === pillbugId) continue;
      if (state.stackHeightAt(targetPos) > 1) continue; // only a lone piece on the ground can be thrown
      if (state.isFrozen(targetId)) continue;
      for (const dest of emptyDests) {
        if (throwKeepsHiveConnected(state, targetId, dest)) {
          moves.push({ kind: MoveKind.THROW, pieceId: pillbugId, to: dest, thrownPieceId: targetId });
        }
      }
    }
  }
  return moves;
}

function throwKeepsHiveConnected(state: GameState, targetId: number, dest: Pos): boolean {
  const targetPos = state.position.get(targetId)!;
  const occ = new Set(state.occupied());
  occ.delete(posKey(targetPos)); // verified height 1 by the caller
  if (!positionsConnected(occ)) return false;
  occ.add(posKey(dest));
  return positionsConnected(occ);
}

File: src/hive_bot/engine/moves.py
"""Legal move generation for the fast Hive engine.

Ports the same rules as tests/reference/helpers.py (slide BFS with the
wedge/gate check, hive connectivity, grasshopper jump, beetle/mosquito
climbing, ladybug's fixed 3-step climb-climb-drop, pillbug throw, turn
structure rules), but computes each piece's full set of legal destinations
in one pass instead of validating one candidate destination at a time --
see `_slide_reachable`. Validated against the reference oracle in
tests/test_engine_vs_reference.py.

Placement actions are generated per piece *type*, not per specific piece
instance: two in-hand pieces of the same type are provably interchangeable
(same owner, same rules, no position yet), so instancing them as separate
actions would only bloat the action space without adding real choices.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from enum import IntEnum

from .constants import HEX_DIRS, TURN_NUMBER_QUEEN_MUST_BE_PLACED, PieceType
from .state import GameState, Pos, neighbors, positions_connected, shared_neighbors


class MoveKind(IntEnum):
    PLACE = 0
    MOVE = 1
    THROW = 2


@dataclass(frozen=True, slots=True)
class Move:
    kind: MoveKind
    piece_id: int  # placed/moved piece, or the pillbug itself for THROW
    to: Pos
    thrown_piece_id: int | None = None  # THROW only: the piece being lifted


def generate_legal_moves(state: GameState) -> list[Move]:
    if state.game_over:
        return []
    moves = _placement_moves(state)
    if state.queen_placed[state.current_player]:
        moves += _board_moves(state)
        moves += _pillbug_throw_moves(state)
    return moves


# --- placement -----------------------------------------------------------


def _placement_moves(state: GameState) -> list[Move]:
    owner = state.current_player
    hand = state.hand[owner]
    if not hand:
        return []
    positions = _placement_positions(state, owner)
    if not positions:
        return []
    must_place_queen = (
        not state.queen_placed[owner] and state.turn_no == TURN_NUMBER_QUEEN_MUST_BE_PLACED
    )
    moves: list[Move] = []
    seen_types: set[PieceType] = set()
    for piece_id in hand:
        piece_type = state.pieces[piece_id].piece_type
        if piece_type in seen_types:
            continue
        if must_place_queen and piece_type != PieceType.QUEEN:
            continue
        seen_types.add(piece_type)
        moves.extend(Move(MoveKind.PLACE, piece_id, pos) for pos in positions)
    return moves


def _placement_positions(state: GameState, owner: int) -> set[Pos]:
    occupied = state.occupied()
    if not occupied:
        return {(0, 0, 0)}

    opponent = 1 - owner
    if state.turn_no == 1:
        # Each side's very first placement just needs to touch the hive so
        # far, which (on turn 1) can only be the opponent's lone piece.
        anchors = state.board_positions(opponent)
        return {nb for pos in anchors for nb in neighbors(pos) if nb not in occupied}

    own_positions = state.board_positions(owner)
    opp_positions = state.board_positions(opponent)
    candidates = {
        nb for pos in own_positions for nb in neighbors(pos) if nb not in occupied
    }
    return {c for c in candidates if all(nb not in opp_positions for nb in neighbors(c))}


# --- on-board moves --------------------------------------------------------


def _board_moves(state: GameState) -> list[Move]:
    owner = state.current_player
    moves: list[Move] = []
    for piece_id in state.pieces_on_board(owner):
        pos = state.position[piece_id]
        assert pos is not None
        if state.top_piece_at(pos) != piece_id:
            continue  # buried under a beetle/mosquito, can't move
        piece_type = state.pieces[piece_id].piece_type
        for dest in _destinations_for(state, piece_id, piece_type):
            if _move_keeps_hive_connected(state, piece_id, dest):
                moves.append(Move(MoveKind.MOVE, piece_id, dest))
    return moves


def _destinations_for(state: GameState, piece_id: int, piece_type: PieceType) -> set[Pos]:
    if piece_type in (PieceType.QUEEN, PieceType.PILLBUG):
        return _slide_destinations(state, piece_id, max_steps=1)
    if piece_type == PieceType.ANT:
        return _slide_destinations(state, piece_id, max_steps=None)
    if piece_type == PieceType.SPIDER:
        return _slide_destinations(state, piece_id, max_steps=3, exact_steps=3)
    if piece_type == PieceType.GRASSHOPPER:
        return _grasshopper_destinations(state, piece_id)
    if piece_type == PieceType.BEETLE:
        return _beetle_destinations(state, piece_id)
    if piece_type == PieceType.LADYBUG:
        return _ladybug_destinations(state, piece_id)
    if piece_type == PieceType.MOSQUITO:
        return _mosquito_destinations(state, piece_id)
    raise AssertionError(f"unhandled piece type {piece_type!r}")


def _resulting_occupied(state: GameState, piece_id: int, dest: Pos) -> frozenset[Pos]:
    """Occupied-position set after hypothetically moving `piece_id` to
    `dest`. Only vacates the start cell if this piece was its sole
    occupant -- a piece stacked under a beetle/mosquito keeps that position
    "occupied" for connectivity purposes when the top piece leaves.
    """
    start = state.position[piece_id]
    assert start is not None
    occ = set(state.occupied())
    if state.stack_height_at(start) == 1:
        occ.discard(start)
    occ.add(dest)
    return frozenset(occ)


def _move_keeps_hive_connected(state: GameState, piece_id: int, dest: Pos) -> bool:
    """The official "One Hive Rule" requires connectivity to hold both
    *during* a move (the instant the piece is lifted, before it lands --
    real Hive rules forbid a move that would momentarily split the hive,
    even if the piece's destination would reconnect it) and *after* it
    (the final board). Only matters "during" when the start cell truly
    empties (stack height 1); a beetle/mosquito climbing off a taller stack
    never affects connectivity by leaving, since something remains there.

    tests/reference/oracle.py's `_board_move_keeps_hive_connected` /
    `_after_move_hive_connected` mirror this same two-check structure
    (corrected from the reference oracle's original, which applied the
    "during" check unconditionally even to stacked-piece departures that
    don't actually vacate their cell -- a real, reachable bug in the
    shipped Django app; see test_engine_vs_reference.py)."""
    start = state.position[piece_id]
    assert start is not None
    if state.stack_height_at(start) == 1:
        during = set(state.occupied())
        during.discard(start)
        if not positions_connected(frozenset(during)):
            return False
    return positions_connected(_resulting_occupied(state, piece_id, dest))


def _slide_destinations(
    state: GameState,
    piece_id: int,
    max_steps: int | None,
    exact_steps: int | None = None,
) -> set[Pos]:
    start = state.position[piece_id]
    assert start is not None
    occupied_without_start = state.occupied() - {start}
    return _slide_reachable(occupied_without_start, start, max_steps, exact_steps)


def _slide_reachable(
    occupied: frozenset[Pos],
    start: Pos,
    max_steps: int | None,
    exact_steps: int | None,
) -> set[Pos]:
    """All destinations reachable from `start` by sliding across `occupied`
    (which must NOT include `start` itself -- the piece is lifted first).
    Exact-step sliding (spider) needs a different search than
    unbounded/max-step sliding (ant/queen) -- see `_slide_reachable_exact`
    for why -- so this just dispatches to whichever is correct.
    """
    if exact_steps is not None:
        return _slide_reachable_exact(occupied, start, exact_steps)

    # A single BFS pass computes every reachable destination instead of
    # validating one candidate at a time: since the wedge/gate check and
    # the "hive stays connected" check at each step only depend on the
    # current node and the single candidate next node (not on the path
    # taken to get there), and *any* non-self-crossing path to a node
    # within max_steps is equally good enough (there's no "must be exactly
    # this many steps" requirement here), a node's reachability doesn't
    # depend on which path first discovered it -- so a single shared
    # `visited` set across the whole search is exact, not an
    # approximation. (This reasoning does NOT extend to exact-step
    # sliding -- see `_slide_reachable_exact`.)
    visited = {start}
    reachable: set[Pos] = set()
    queue: deque[tuple[Pos, int]] = deque([(start, 0)])
    while queue:
        current, steps = queue.popleft()
        for nb in neighbors(current):
            if nb in visited:
                continue
            visited.add(nb)
            if nb in occupied:
                continue
            if not positions_connected(occupied | {nb}):
                continue
            n1, n2 = shared_neighbors(current, nb)
            if n1 in occupied and n2 in occupied:
                continue  # wedged: both flanking hexes occupied

            new_steps = steps + 1
            if max_steps is None or new_steps <= max_steps:
                reachable.add(nb)
            if max_steps is None or new_steps < max_steps:
                queue.append((nb, new_steps))
    return reachable


def _slide_reachable_exact(
    occupied: frozenset[Pos], start: Pos, exact_steps: int
) -> set[Pos]:
    """Destinations reachable from `start` in *exactly* `exact_steps`
    slides (spider), without ever crossing a hex already visited earlier
    in that same slide.

    Unlike max-step sliding, this can't reuse one shared `visited` set
    across the whole search: a hex first reached via a *shorter* path
    would then block rediscovering it via a *different*, longer,
    non-self-crossing path that legitimately reaches it in exactly
    `exact_steps` -- "no revisiting a hex" is a per-path rule, not a
    global one across every path from `start`. A real game found this
    concretely: a destination the shared-visited BFS marked unreachable
    (some other path had already visited it after 2 steps) but which a
    valid 3-step path -- through different intermediate hexes -- does
    reach. So this instead explores every non-self-crossing path
    explicitly; cheap in practice since `exact_steps` is always 3 (only
    spider uses this) with branching factor at most 6.

    This exact bug is present in the vendored reference's own
    `can_slide_path` (tests/reference/helpers.py -- an unmodified copy of
    the real ttbg-web-app's own code, not just this repo's re-transcribed
    oracle), so it isn't just a bug here -- it's a real, reachable bug in
    the shipped Django app itself. test_engine_vs_reference.py's oracle
    comparison inherits it via `tests/reference/oracle.py`, which calls
    the same `can_slide_path` -- see that test file for how spider moves
    are carved out of the direct oracle comparison as a result.
    """
    reachable: set[Pos] = set()

    def dfs(current: Pos, path: frozenset[Pos], steps: int) -> None:
        if steps == exact_steps:
            reachable.add(current)
            return
        for nb in neighbors(current):
            if nb in path or nb in occupied:
                continue
            if not positions_connected(occupied | {nb}):
                continue
            n1, n2 = shared_neighbors(current, nb)
            if n1 in occupied and n2 in occupied:
                continue  # wedged: both flanking hexes occupied
            dfs(nb, path | {nb}, steps + 1)

    dfs(start, frozenset({start}), 0)
    return reachable


def _grasshopper_destinations(state: GameState, piece_id: int) -> set[Pos]:
    start = state.position[piece_id]
    assert start is not None
    occupied = state.occupied()
    destinations: set[Pos] = set()
    for dq, dr, ds in HEX_DIRS:
        q, r, s = start
        hopped_over = 0
        while True:
            q, r, s = q + dq, r + dr, s + ds
            pos = (q, r, s)
            if pos not in occupied:
                if hopped_over > 0:
                    destinations.add(pos)
                break
            hopped_over += 1
    return destinations


def _beetle_destinations(state: GameState, piece_id: int) -> set[Pos]:
    start = state.position[piece_id]
    assert start is not None
    piece_level = state.stack_index_of(piece_id)
    destinations: set[Pos] = set()
    for end in neighbors(start):
        end_stack_len = state.stack_height_at(end)
        if piece_level != end_stack_len:
            # Climbing on/off a stack always changes level -- no gate check.
            destinations.add(end)
            continue
        n1, n2 = shared_neighbors(start, end)
        h1 = state.stack_height_at(n1) - 1  # top-of-stack index, -1 if empty
        h2 = state.stack_height_at(n2) - 1
        if h1 >= piece_level and h2 >= piece_level:
            continue  # wedged between two taller (or equal) neighbors
        destinations.add(end)
    return destinations


def _ladybug_destinations(state: GameState, piece_id: int) -> set[Pos]:
    start = state.position[piece_id]
    assert start is not None
    occupied = set(state.occupied())
    occupied.discard(start)
    destinations: set[Pos] = set()
    for a in neighbors(start):
        if a not in occupied:
            continue
        for b in neighbors(a):
            if b == start or b not in occupied:
                continue
            for c in neighbors(b):
                if c in (start, a) or c in occupied:
                    continue
                destinations.add(c)
    return destinations


def _mosquito_destinations(state: GameState, piece_id: int) -> set[Pos]:
    if state.stack_index_of(piece_id) > 0:
        # Having climbed up (by copying a beetle) it now only moves as one.
        return _beetle_destinations(state, piece_id)
    start = state.position[piece_id]
    assert start is not None
    adjacent_types = {
        state.pieces[top].piece_type
        for nb in neighbors(start)
        if (top := state.top_piece_at(nb)) is not None
    }
    adjacent_types.discard(PieceType.MOSQUITO)
    destinations: set[Pos] = set()
    for piece_type in adjacent_types:
        destinations |= _destinations_for(state, piece_id, piece_type)
    return destinations


# --- pillbug throw ---------------------------------------------------------


def _pillbug_throw_moves(state: GameState) -> list[Move]:
    owner = state.current_player
    moves: list[Move] = []
    for pillbug_id in state.pieces_on_board(owner):
        if state.pieces[pillbug_id].piece_type != PieceType.PILLBUG:
            continue
        pos = state.position[pillbug_id]
        assert pos is not None
        if state.top_piece_at(pos) != pillbug_id:
            continue
        occupied = state.occupied()
        empty_dests = [nb for nb in neighbors(pos) if nb not in occupied]
        if not empty_dests:
            continue
        for target_pos in neighbors(pos):
            target_id = state.top_piece_at(target_pos)
            if target_id is None or target_id == pillbug_id:
                continue
            if state.stack_height_at(target_pos) > 1:
                continue  # only a lone piece on the ground can be thrown
            if state.is_frozen(target_id):
                continue
            for dest in empty_dests:
                if _throw_keeps_hive_connected(state, target_id, dest):
                    moves.append(
                        Move(MoveKind.THROW, pillbug_id, dest, thrown_piece_id=target_id)
                    )
    return moves


def _throw_keeps_hive_connected(state: GameState, target_id: int, dest: Pos) -> bool:
    target_pos = state.position[target_id]
    assert target_pos is not None
    occ = set(state.occupied())
    occ.discard(target_pos)  # verified height 1 by the caller
    if not positions_connected(frozenset(occ)):
        return False
    occ.add(dest)
    return positions_connected(frozenset(occ))
